### ISSUE TITLE
Add ReportDiagnostic API to all initial transform APIs

### ISSUE THREAD

**Issue by Sergio0694** [dotnet/roslyn#63776]
## Background and Motivation

Right now, one of the main pain points with incremental generators is how to produce and emit diagnostics from a transform node. What generator authors tend to do (eg. the regex generator in the runtime does this, I do the same in the generators for both the MVVM Toolkit and ComputeSharp, etc.) is to do something like this:

```txt
- Inputs (eg. CreateSyntaxProvider)
  |- (Domain models, Diagnostics)
    |- Select models -> Add comparers -> Produce source
    |- Select diagnostics -> Produce source just to report them
```

This has multiple downsides:

- It's unintuitive
- It's very clunky, as you now have to build a diagnostics collection, pass that down, then filter it, etc.
- It's rather inefficient, as you're adding more pipeline steps just to split models and diagnostics and emit them
- Diagnostics can root (even accidentally) trees and symbols, so having them in incremental producers isn't great

This is a pain point that's been brought up in the context of incremental source generators in the past by others as well.

## Proposed API

The proposed API assumes that the initial transform step would produce domain models directly, so that one would never have symbols in incremental value(s) providers. As such, given all diagnostics info will be gathered together with the domain models, we only need the ability to report diagnostics in the same place where this initial transformation is done. Currently, that'd be:

```diff
 namespace Microsoft.CodeAnalysis;

public readonly struct GeneratorAttributeSyntaxContext
{
+   public void ReportDiagnostic(Diagnostic diagnostic);
}

public readonly struct GeneratorSyntaxContext
{
+   public void ReportDiagnostic(Diagnostic diagnostic);
}
```

### BEFORE

``` csharp
// Gather info for all annotated command methods
IncrementalValuesProvider<(Model, ImmutableArray<Diagnostic>)> modelsAndDiagnostics =
    context.SyntaxProvider
    .CreateSyntaxProvider(
        static (node, _) => node is MethodDeclarationSyntax,
        static (context, _) => GetModelAndProduceDiagnostics(context.Node, context.SemanticModel));

IncrementalValuesProvider<Model> models = modelsAndDiagnostics.Select(pair => pair.Item1);
IncrementalValuesProvider<ImmutableArray<Diagnostic>> diagnostics = modelsAndDiagnostics.Select(pair => pair.Item2);

// Output the diagnostics
context.RegisterSourceOutput(diagnostics, static (context, diagnostics) =>
{
    foreach (Diagnostic diagnostic in diagnostics)
    {
        context.ReportDiagnostic(diagnostic);
    }
});

// Generate the commands
context.RegisterSourceOutput(models, static (context, item) =>
{
    CompilationUnitSyntax compilationUnit = GetTheProducedSyntax(item);

    context.AddSource($"whatever.g.cs", compilationUnit.GetText(Encoding.UTF8));
});
```

### AFTER

```csharp
// Gather info for all annotated command methods
IncrementalValuesProvider<Model> models =
    context.SyntaxProvider
    .CreateSyntaxProvider(
        static (node, _) => node is MethodDeclarationSyntax,
        static (context, _) => GetModelAndProduceDiagnostics(context));

// Generate the commands
context.RegisterSourceOutput(models, static (context, item) =>
{
    CompilationUnitSyntax compilationUnit = GetTheProducedSyntax(item);

    context.AddSource($"whatever.g.cs", compilationUnit.GetText(Encoding.UTF8));
});
```

## Alternative Designs

None, just keep doing what we're doing, which has all the downsides mentioned above.

## Risks

None, other than just the additional complexity of two more APIs. It seems to me the current design is pushing devs into creating worse and more inefficient generators anyway, so I'd say if anything this proposal would reduce risk once implemented.

**Comment by CyrusNajmabadi** [comment]
This concept seems very reasonable to me (specifically allowing diagnostics at any point,a nd trackign those forward to the end).  @chsienki do you want to bring to API review?

**Comment by Sergio0694** [comment]
I've updated the proposal with APIs to also support emitting diagnostics from a `Select` step, which would be useful in several scenarios where generator authors currently also have to manually track diagnostics, emit them, and then have separate filtering steps to branch them off and route them to a dedicated source production node just to emit them, which is not great both for usability and for performance. The proposed API shape fully addresses that scenario as well 🙂

**Comment by Sergio0694** [comment]
Sharing updates here following a conversation with @CyrusNajmabadi on Discord. The regex generator (cc. @stephentoub) is also handling diagnostics incorrectly ([source here](https://github.com/dotnet/runtime/blob/db4a954e412b09dd0b32ba482900de45936e268d/src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.cs#L46)), in that it's producing `Diagnostic` instances, which internally track a `Location` retrieved from a symbol, and then branching them off to a source production node to emit them. These `Diagnostic` objects are not equatable, which breaks incrementality, which is the point Cyrus brought up as being bad. So it seems at the very least this generator from the SDK could also use a helper API like the one proposed here.

Pretty sure the COM source generator could use this as well. Same goes, as mentioned in the proposal, for all the generators in the MVVM Toolkit as well (which currently are doing the same inefficient thing as the regex generator). Just to name a few.

**Comment by CyrusNajmabadi** [comment]
> These Diagnostic objects are not equatable,

Diagnostics seem to be `IEquatable<Diagnostic?>`.

**Comment by Sergio0694** [comment]
By "not equatable" I didn't mean that they don't implement the interface, but that they'd return `false`. Unless I'm reading things wrong, if you get a `Location` from a symbol (which is what it's being done currently), that `Location` object will internally track the original syntax tree, which will not match across re-runs of the generator, right? 🤔

**Comment by CyrusNajmabadi** [comment]
@Sergio0694 that would be correct.  hopefully these diagnostics are rarely used.

**Comment by jkoritzinsky** [comment]
Aren't `SyntaxTree` objects reference-equatable between `Compilation` instances if they're unchanged? I could have sworn I validated that previously. If they're reference-equatable, then the `Location` instances would also compare equal.

**Comment by CyrusNajmabadi** [comment]
the issue is when tehy're not the same.  e.g. any edits to teh file at all will always need us to rerun every potential path that might report a diagnostic, as the diagnostics may now be in a different location.  Would def suggest using analyzers here instead.

**Comment by Sergio0694** [comment]
> "any edits to teh file at all will always need us to rerun every potential path that might report a diagnostic"

I still don't understand why that would be an issue for the ones passed to `ReportDiagnostic` from `GeneratorAttributeSyntaxContext` or `GeneratorSyntaxContext`. Those two providers give you symbols, meaning they would already re-run for every edit regardless, no? So I don't see how there would be an issue with diagnostics causing more re-runs, wouldn't that source node still re-run anyway?

I can see how this would not be the case for ones passed to an arbitrary `Select`, but at least for the proposed `ReportDiagnostic` in those two root source providers, wouldn't that be ok and not causing additional re-runs? 🤔

**Comment by jkoritzinsky** [comment]
> the issue is when tehy're not the same. e.g. any edits to teh file at all will always need us to rerun every potential path that might report a diagnostic, as the diagnostics may now be in a different location. Would def suggest using analyzers here instead.

Yes, and in those cases you might need to run because the diagnostics might change (such as the issue being fixed and a diagnostic no longer being required). Unless your suggestion is to never emit diagnostics from generators and duplicate all of the logic into an analyzer (which is the exact opposite of the guidance we on the interop team were given a while back).

**Comment by CyrusNajmabadi** [comment]
>  and duplicate all of the logic into an analyzer (which is the exact opposite of the guidance we on the interop team were given a while back).

Yes.  This would be my guidance.  This also means that later parts of the compilation path are not gated on these diagnostics being computed.  e.g. they can be computed asynchronously (or outside of VS) etc. etc. etc.

**Comment by jkoritzinsky** [comment]
Any particular reason the guidance changed here?

Also, if this is the guidance, we should deprecate all APIs that allow reporting diagnostics in a source generator. Otherwise developers will take the easy direction to report diagnostics in a generator and will fall into a performance trap.

Alternatively, could we determine a better way to trigger generators such that we don't need to duplicate our logic everywhere?

**Comment by CyrusNajmabadi** [comment]
> Any particular reason the guidance changed here?

Performance.  Things are pretty bad right now :)  Getting things out of the main generation path is good.

--

Note: i don't know if official guidance has changed.  But this is definitely my observation having to go fix an enormous number of perf issues in generators that are a big impact on everything above.  My preference is that everything non-essential get out of the generator path.

--

> Also, if this is the guidance, we should deprecate all APIs that allow reporting diagnostics in a source generator. 

I agree with this.  @chsienki can we do that?

**Comment by CyrusNajmabadi** [comment]
> Alternatively, could we determine a better way to trigger generators such that we don't need to duplicate our logic everywhere?

Could you clarify what you mean by that?

**Comment by stephentoub** [comment]
>> we should deprecate all APIs that allow reporting diagnostics in a source generator.

> I agree with this. @chsienki can we do that?

I don't understand this. If the generator can't generate code because of an issue with the input, how is it supposed to communicate that?

Duplicating everything into a separate analyzer sounds terrible.

**Comment by stephentoub** [comment]
Also, if diagnostucs are only issued about error conditions, is there really a perf problem?

**Comment by CyrusNajmabadi** [comment]
> I don't understand this. If the generator can't generate code because of an issue with the input, how is it supposed to communicate that?

It coudl do so in an anlayzer.

> Duplicating everything into a separate analyzer sounds terrible.

Presumably they would just share a helper here.  It would just be about where/when we run this.  Allowing it to happen asynchronously without blocking comptuation generation.

**Comment by CyrusNajmabadi** [comment]
> Also, if diagnostucs are only issued about error conditions, is there really a perf problem?

Definitely.  The problem is that any edit potentially invalidates all diagnostics.  E.g. any edit may change teh positions of anything.  So that makes "positions of all text" part of the model that all generators implicitly now have.  So any edit must rerun everything :-/   That's not... ideal.

**Comment by stephentoub** [comment]
> The problem is that any edit potentially invalidates all diagnostics

Sure, but that would only trigger a generator if it had output diagnostics. And if it only does so on error, I'm questioning how impactful that really is.

**Comment by stephentoub** [comment]
> Presumably they would just share a helper here.

Even if they do, it's poor from a usability perspective IMO. I'm writing my generator, I encounter bad input, and rather then just emitting a diagnostic, i need to go off and refactor my code to be able to detect and handle it in a completely different component.

**Comment by CyrusNajmabadi** [comment]
> Sure, but that would only trigger a generator if it had output diagnostics. And if it only does so on error, I'm questioning how impactful that really is.

From my perspective, errors are the normal case :)  e.g. it's expected to commonly have errors the majority of hte time while editing.  if your particular use case is that this is extremely rare, it's possibly ok.

**Comment by CyrusNajmabadi** [comment]
> Even if they do, it's poor from a usability perspective IMO

I mean... usability is like my 10th level concern here.  These are expert apis intended for extremely advanced scenarios.  And they have a deleterious impact on performance.  I think we should focus on what can be blazing fast, with everything unnecessary moved outward, and then care about things like usability much further down the line :)

**Comment by stephentoub** [comment]
If no one can write them, I guess we won't have any performance problems.

**Comment by stephentoub** [comment]
(When they are written, though, we will have new problems: folks will write generators, and then because they can't emit diagnostics, they just won't.  Now we'll have a bunch of totally undebuggable source generators that impact end-user usability.)

**Comment by CyrusNajmabadi** [comment]
> If no one can write them, I guess we won't have any performance problems.

Generally speaking, i actually agree with that.  I think putting these out as a generalized solution (with their current designs) was somewhat of a mistake.  We should have made this internal, expert-only, and only published when we had something suitable that we could know could fit into a "this is required to run" and "this must not be a perf issue" framework.

We're in a bad state now because we have the "this is required to run" part, but also "perf is really bad" as well.  It's not a good place to be :-/

> (When they are written, though, we will have new problems: folks will write generators, and then because they can't emit diagnostics, they just won't. Now we'll have a bunch of totally undebuggable source generators that impact end-user usability.)

I get the usability issue.  I do.  But we also have hte issue of customers reporting bad performance issues because of generators as well.  So i'm trying to think about what can be done about that that maintains a particualr user experience for end customers, even if it is more work for us on the authoring side.

**Comment by CyrusNajmabadi** [comment]
Note: talking to @jkortech offline, you guys may be ok thanks to the use of ForAttributeWithMetadataName making it so that these issues are already getting a several-orders-of-magnitude boost.  What i'm referring to here is the request that these APIs be pushed earlier in the pipeline, where they will have more of a perf impact.

**Comment by stephentoub** [comment]
> So i'm trying to think about what can be done about that that maintains a particualr user experience for end customers

And that's great; we obviously both want great experiences all around.  I'd prefer it if we not just jump to throwing the baby out with the bathwater, though.  Surely there must be other options still to be considered.

**Comment by CyrusNajmabadi** [comment]
> Surely there must be other options still to be considered.

Sorry for the confusion.  

I had a long conversation with RefSergio about thsi over the weekend.  My recommendation is to do what you guys (SDK) are *already* doing in your approach.  Specifically, you are making the diagnostic part of your data model.  Your approach *is* teh one i am suggesting as acceptable as it clearly encodes into the value-tracking-system we have the information necessary which effects the outcome of the generator (both source and diagnostics produced).

My recommendation is that if that is not something the author is willing to do (which i personally think is reasonable) then they switch to using analyzer to report the diagnostics and they do not have diagnostics in their pipeline flow at all.

**Comment by jkoritzinsky** [comment]
So, looking at this API again, I think we could actually get decent performance out of it without having the issues we've mentioned around `Location` objects.

For each of the overloads (especially the `Select` and `SelectMany` ones), the generator would internally produce a regular non-output node that would calculate the mapped result, but it would not save the diagnostics. It would also produce an "output" node that would collect the diagnostics, but not the mapped result. Then we wouldn't be caching `Diagnostic` instances or `Location` instances, because we wouldn't record them in the tables. We'd only collect them for the output nodes (same as how the existing output nodes do not cache any output, and always recalculate the last step). For generators that only emit diagnostics in the same step that also creates the internal data model, this wouldn't result in any steps running more than necessary.

**Comment by Sergio0694** [comment]
> "I had a long conversation with RefSergio"

Stephen's not on Discord so he'll probably not recognize the nickname, anyway that'd be me 😄

> "My recommendation is to do what you guys (SDK) are already doing in your approach. Specifically, you are making the diagnostic part of your data model."

I mean I'm glad to hear that that's acceptable, but that's not what I understood from our conversation. That branching is what I'm _already_ doing in my generators, but you still said that's very inefficient and that the ideal solution would actually be to split everything into a generator and a separate analyzer. So I guess I'm a bit confused now. If the branching is indeed ok though, then I'd keep saying that the current proposal would be a nice addition because it'd make such a pattern much more user friendly and less error prone, and would nudge more generator authors towards "doing the right thing".
"Pit of success" like Stephen says and all 🙂

> "For each of the overloads (especially the Select and SelectMany ones), the generator would internally produce a regular non-output node [...]"

That's overall what my original idea was as well. It's effectively be the same semantics as what people are doing manually today, but without the extra clunkyness of having to manually build an array of diagnostics, passing it through, branching twice, selecting the diagnostics, producing them, selecting the other item with the model, push it down the pipeline...

**Comment by b-straub** [comment]
> > and duplicate all of the logic into an analyzer (which is the exact opposite of the guidance we on the interop team were given a while back).
> 
> Yes. This would be my guidance. This also means that later parts of the compilation path are not gated on these diagnostics being computed. e.g. they can be computed asynchronously (or outside of VS) etc. etc. etc.

Can you please advise how to solve the duplicate diagnostic problem for Intellisense?
For my SG I have analyzers showing potential problems breaking the source generation along with code fixes to solve those problems. Now I have two possible scenarios but both are not satisfying:
1) Use the analyzers and report the diagnostics also from SourceProductionContext
2) Use only the analyzers

Problems with:
1) will show duplicate Intellisense entries likely because not equitable even pointing to same location and symbol 
2) no diagnostics in build log and thus not visible if "Build only" chosen for error panel

**Comment by Sergio0694** [comment]
Hey @chsienki, just touching base on to double check whether there's any other points that need resolving or things I should add to the proposal, or if this is good enough to bring it to API review? I'm really looking forward to this one 😄
Not super familiar with the Roslyn proposal flow so wondering if this needed some "ready for review" tag or anything.
Thank you!

**Comment by 333fred** [comment]
### API Review

* Are the `Select` methods a source-breaking change for lambdas?
    * Maybe not, exist Selects are two-args?
    * Not a source-breaking change
* Do we want `Where` as well?
* Are diagnostics a side-channel?
* How would we actually implement this? We'd have a lot of the same issues of tracking and invalidating those locations
    * We could do it the same way users can already, but that's no more efficient
        * Slight optimizations possible, but no meaningful optimizations
    * Possible for more efficient impl, but it's a very big architecture change.
* The pattern we've seen so far has been a separate analyzer
    * Lots of duplication between analyzer and generator
    * Analyzers can be disabled
* These extensions could be written as a 3rd party without needing to be in Roslyn.
    * Could do it as extensions for now and collect info on how expensive these things are?

Conclusion: We will start with 3rd party extension methods and see what data we get about how expensive these are.
